staging/src/k8s.io/apimachinery/third_party/forked/golang/json/fields.go

```diff
@@ -35,6 +35,9 @@ func LookupPatchMetadata(t reflect.Type, jsonField string) (
 		elemType = t.Elem()
 		return
 	}
+	if t.Kind() == reflect.Slice || t.Kind() == reflect.Array {
```

Comment:
why was this change needed? I'm not seeing additional calls to LookupPatchMetadata that would have failed before, and this seems to be descending an additional level in a way inconsistent with how maps behave (previous if case)

Reply:
The reason is here: https://github.com/kubernetes/kubernetes/pull/51321/files/3076f7b4eddad0b1a94dce511bbc1a57b7cfa06c#diff-101008d96c4444a5813f7cb6b54aaff6L1607

It used to call `fieldType.Elem()`. But to make openapi works here, I decide to move the logic of how to get a element type of a slice a level down.

Reply:
How openapi get the elem of an slice is handled here: 
https://github.com/kubernetes/kubernetes/pull/51321/files#diff-624b60a2601b30204424723e9dab6300R72

The difference between openapi and fromStruct makes me decide to move the logic a level down.

Reply:
the treatment of slice and map here still appears oddly inconsistent... a slice returns the element type and metadata from the named field in the element type. a map returns the element type but no metadata for the named field?

Reply:
I guess you misread the code somehow.

`t.Kind() == reflect.Map` handles a primitive map which doesn't support mergekey and patchstrategy.
`t.Kind() == reflect.Slice || t.Kind() == reflect.Array` handles slices.
`t.Kind() != reflect.Struct` handles an object struct, which is a map in json.

Does it make more sense to you now?

Reply:
No, it means the type passed in can mean three different things, which is extremely hard to follow. Prior to this change, the type arg meant two different things. This change makes it worse.

Reply:
Another option is revert the in this file (fields.go). And add another func in the interface:
```
type LookupPatchMeta interface {
	// Lookup the patch metadata (patch strategy and merge key)
	LookupPatchMetadata(key string) (LookupPatchMeta, *PatchMeta, error)
	// Get the type name of the field
	Name() string
        // Use element of a list or a map
        Elem()  // New
}
```

There was one revision of this PR using this interface, I can change it back.

Reply:
After recalling why I didn't choose to have `Elem()` in `LookupPatchMeta interface` and re-evaluating the 2 options, I still think it's better to do it using the current implementation (i.e. no `Elem()` in the interface).

> it means the type passed in can mean three different things, which is extremely hard to follow. Prior to this change, the type arg meant two different things. This change makes it worse.

Admittedly supporting slice here will introduce some complexity, but IMO it will simplified the overall complexity by centralize all of the logic of handling how to get the sub-element or sub-type. 

If you insist not doing this way, I can address this comment in a followup PR. Because code freeze is about 2 weeks away, I'd like to merge the major logic ASAP to have some time to discover potential issue. Otherwise it may be postponed like last release.

Reply:
> Admittedly supporting slice here will introduce some complexity, but IMO it will simplified the overall complexity by centralize all of the logic of handling how to get the sub-element or sub-type.

It's not just complexity, it makes the interface method mean different things depending on what is passed in. That's not a good interface. This are of code is already extremely difficult to understand, I don't think we should confuse the interfaces further.

Reply:
> It's not just complexity, it makes the interface method mean different things depending on what is passed in.

Are you saying it sometimes returns nil `*PatchMeta` and sometimes non-nil?

> That's not a good interface.

What the good interface you imagine for this case?

Reply:
> Are you saying it sometimes returns nil *PatchMeta and sometimes non-nil?

As written in this pull, sometimes it descends two levels (for maps and slices), sometimes one (for other field types). Sometimes it returns patch/schema info (for slices and non-map fields), sometimes it doesn't (for maps).

These don't actually encapsulate complexity, since the caller still has to know to merge map and slice fields differently, but it does mean that the method can be called incorrectly and the results used incorrectly accidentally

> What the good interface you imagine for this case?

Well, we have three distinct behaviors that callers depend on, so I'd probably expect three methods:
* One for getting the schema of the element of a map field, fails if called on a non-map field
* One for getting the schema of the element of a slice field and the patch metadata for the slice, fails if called on a non-slice
* One for getting schema and patch metadata for a struct field

I'm not envisioning anything particularly fancy, just breaking apart the three cases the current method handles differently into distinct methods to protect from accidentally calling the wrong way

Reply:
Thanks for your explanation.
I will try to do it that way.